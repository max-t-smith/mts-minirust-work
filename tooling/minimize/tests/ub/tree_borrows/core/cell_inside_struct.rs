//@ compile-flags: --minimize-tree-borrows

use std::cell::Cell;

struct Foo {
    field1: u32,
    field2: Cell<u32>,
}

pub fn main() {
    let root = Foo { field1: 42, field2: Cell::new(88) };

    unsafe {
        let a = &root;
        let a: *mut Foo = a as *const _ as *mut _;

        // Writing to `field2`, which is interior mutable, should be allowed.
        (*a).field2.set(10);

        // Writing to `field1`, which is frozen, should not be allowed.
        (*a).field1 = 88; // UB! Child write to frozen.
    }
}
