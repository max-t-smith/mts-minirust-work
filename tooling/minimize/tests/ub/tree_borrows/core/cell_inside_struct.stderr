fatal error: UB: Tree Borrows: writing to the local of a pointer with Frozen permission
